src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py

```diff
@@ -0,0 +1,130 @@
+    refer to this superclass for more information regarding those methods.
+
+    Args:
+        vocab_file (`str`):
+            Path to the vocabulary file.
+        merges_file (`str`):
+            Path to the merges file.
+        errors (`str`, *optional*, defaults to `"replace"`):
+            Paradigm to follow when decoding bytes to UTF-8. See
+            [bytes.decode](https://docs.python.org/3/library/stdtypes.html#bytes.decode) for more information.
+        unk_token (`str`, *optional*, defaults to `<|endoftext|>`):
+            The unknown token. A token that is not in the vocabulary cannot be converted to an ID and is set to be this
+            token instead.
+        bos_token (`str`, *optional*, defaults to `<|endoftext|>`):
+            The beginning of sequence token.
+        eos_token (`str`, *optional*, defaults to `<|endoftext|>`):
+            The end of sequence token.
+        add_prefix_space (`bool`, *optional*, defaults to `False`):
+            Whether or not to add an initial space to the input. This allows to treat the leading word just as any
+            other word. (GPTNeoX tokenizer detect beginning of words by the preceding space).
+        trim_offsets (`bool`, *optional*, defaults to `True`):
+            Whether or not the post-processing step should trim offsets to avoid including whitespaces.
+    """
+
+    vocab_files_names = VOCAB_FILES_NAMES
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+    max_model_input_sizes = PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
+    model_input_names = ["input_ids", "attention_mask"]
+    slow_tokenizer_class = GPTNeoXTokenizer
+
+    def __init__(
+        self,
+        vocab_file=None,
+        merges_file=None,
+        tokenizer_file=None,
+        unk_token="<|endoftext|>",
+        bos_token="<|endoftext|>",
+        eos_token="<|endoftext|>",
+        add_prefix_space=False,
+        **kwargs
+    ):
+        super().__init__(
+            vocab_file,
+            merges_file,
+            tokenizer_file=tokenizer_file,
+            unk_token=unk_token,
+            bos_token=bos_token,
+            eos_token=eos_token,
+            add_prefix_space=add_prefix_space,
+            **kwargs,
+        )
+
+        pre_tok_state = json.loads(self.backend_tokenizer.pre_tokenizer.__getstate__())
+        if pre_tok_state.get("add_prefix_space", add_prefix_space) != add_prefix_space:
+            pre_tok_class = getattr(pre_tokenizers, pre_tok_state.pop("type"))
+            pre_tok_state["add_prefix_space"] = add_prefix_space
+            self.backend_tokenizer.pre_tokenizer = pre_tok_class(**pre_tok_state)
+
+        self.add_prefix_space = add_prefix_space
+
+    def save_vocabulary(self, save_directory: str, filename_prefix: Optional[str] = None) -> Tuple[str]:
+        files = self._tokenizer.model.save(save_directory, name=filename_prefix)
+        return tuple(files)
```

Comment:
@SaulLu could you chime in about the changes done here, especially regarding how to mirror these to the slow tokenizer? I wonder if this mean that a new tokenizer class is needed, or if the GPT-2 tokenizer can still be leveraged by using the `add_prefix_space` parameter.

Reply:
With pleasure! :smile: Looking at this fast file, I think your earlier suggestion to directly use the GPT2 tokenizer makes sense!

The only differences I see between `GPT2TokenizerFast` and `GPTNeoXTokenizerFast` are that the `_batch_encode_plus`, `_encode_plus` and `_build_conversation_input_ids`  methods are redefined in `GPT2TokenizerFast` and they are not in `GPTNeoXTokenizerFast`. If it useful, you'll find below a screenshot showing the differences between `GPT2TokenizerFast` (left) and `GPTNeoXTokenizerFast` (right)

![image](https://user-images.githubusercontent.com/55560583/165624366-3936489b-ebf6-4714-8d4c-9c93cc5f593e.png)


In  `_batch_encode_plus` and `_encode_plus` , GPT2TokenizerFast forces `add_prefix_space` to be set to true when the inputs are already pretokenized.  Is there a use case here for which this blocking would not be desirable?

Regarding `_build_conversation_input_ids`, I think it has not been added to `GPTNeoXTokenizerFast` because there is no use case identified but I think it is not inconvenient to have it.

Reply:
I think the because out tokenizer is configured slightly differently, we don't want to enforce `add_prefix_space` (though I'm not 100% sure). I've added the conversation function though.